Enable InvertIf

chborl opened the pull request:
PR to reenable the InvertIf feature.

chborl commented:
This PR is a work in progress.  This feature has been disable for some time so if there are any new language features that you think may impact it please let me know.

Known issues that will be addressed in this PR :
- currently, VB doesn't trigger on the If of an If-ElseIf statement.  I'm updating it to trigger on Ifs and ElseIfs

chborl commented:
@dotnet/roslyn-analysis , @dotnet/roslyn-ide  for review.

jinujoseph commented:
@chborl can you file a issue for VB missing scenario expressed above

jinujoseph commented:
cc @CyrusNajmabadi , @MaStr11

chborl commented:
# 25434, Invertif on VB should work when subsequent block is an elseif

chborl commented:
@dotnet-bot retest windows_release_vs-integration_prtest please

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,643 @@
+            await TestFixOneAsync(
+@"[||]if (a) { a(); } else { b(); }",
+@"        if (!a) { b(); }
+        else { a(); }");
```
seems not good that formatting is changing.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            // statement matches the right if statement after the edit.
+            var newIfNodeStatement = ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement ?
+                SyntaxFactory.Block(ifNode.Else.Statement) :
+                ifNode.Else.Statement;
```
IDE style is:

```c#
ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement
  ? SyntaxFactory.Block(ifNode.Else.Statement)
  : ifNode.Else.Statement;
```

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            var invertedIf = ifNode.WithCondition(Negate(ifNode.Condition, semanticModel, cancellationToken))
+                      .WithStatement(newIfNodeStatement)
+                      .WithElse(ifNode.Else.WithStatement(ifNode.Statement))
+                      .WithAdditionalAnnotations(Formatter.Annotation);
```
formatting seems quite dangerous to just go and explicitly do.  It feels like we shouldn't have to format in general for this refactoring.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            return document.WithSyntaxRoot(result);
+        }
+
+        private bool IsComparisonOfZeroAndSomethingNeverLessThanZero(BinaryExpressionSyntax binaryExpression, SemanticModel semanticModel, CancellationToken cancellationToken)
```
can you comment what this method is trying to do?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            if (binaryExpression.Kind() == SyntaxKind.GreaterThanExpression &&
+                binaryExpression.Right.Kind() == SyntaxKind.NumericLiteralExpression)
+            {
+                canSimplify = CanSimplifyToLengthEqualsZeroExpression(
```
several of these branches seem identical (or just swap left/right), can we make this simpler?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            CancellationToken cancellationToken)
+        {
+            var numericValue = semanticModel.GetConstantValue(numericLiteralExpression, cancellationToken);
+            if (numericValue.HasValue && numericValue.Value is int && (int)numericValue.Value == 0)
```
```numericValue is int val && val == 0```

Neme12 replied:
what about `is 0`?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            {
+                var symbol = semanticModel.GetSymbolInfo(variableExpression, cancellationToken).Symbol;
+
+                if (symbol != null && (symbol.Name == "Length" || symbol.Name == "LongLength"))
```
```nameof(Array.Length)```

chborl replied:
Thanks @CyrusNajmabad , I tried to do the same with "LongLength" but error message 'Array' does not contain a definition for 'LongLength'.  This project targets netstandard1.3 and Array.LongLength isn't available till netstandard2.0.  I'll use a constant instead.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+    internal partial class InvertIfCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        private static readonly Dictionary<SyntaxKind, Tuple<SyntaxKind, SyntaxKind>> s_binaryMap =
+            new Dictionary<SyntaxKind, Tuple<SyntaxKind, SyntaxKind>>(SyntaxFacts.EqualityComparer)
```
Use a normal tuple.  give names to the Item0 and Item1.

alrz replied:
this is fixed in https://github.com/dotnet/roslyn/pull/24649

I don't understand why we are removing/adding this class.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            {
+                var binaryExpression = (BinaryExpressionSyntax)expression;
+                var expressionType = tuple.Item1;
+                var operatorType = tuple.Item2;
```
deconstruct.

CyrusNajmabadi commented:
Done with first pass.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,501 @@
+            cancellationToken As CancellationToken) As Boolean
+
+            Dim value = semanticModel.GetConstantValue(numericLiteralExpression)
+            If value.HasValue AndAlso TypeOf value.Value Is Integer AndAlso DirectCast(value.Value, Integer) = 0 Then
```
this looks duplicated with C# version.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,501 @@
+Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InvertIf
+    <ExportCodeRefactoringProvider(LanguageNames.VisualBasic, Name:=PredefinedCodeRefactoringProviderNames.InvertIf), [Shared]>
+    Friend Class InvertIfCodeRefactoringProvider
+        Inherits CodeRefactoringProvider
```
i'm surprised by the amount of duplicatoin between this and hte C# version. I would have expected a common base class that shared a bunch of the "high level" concepts here, with calls to the subclasses to do the fidly bits where C# and VB differ substantively

chborl replied:
@CyrusNajmabadi , agreed.  But since this was written several years ago and seems to be working well as is, we're opting to not rewrite it at this point.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,501 @@
+            semanticModel As SemanticModel,
+            cancellationToken As CancellationToken) As Boolean
+
+            Dim canSimplify = False
```
why is this temporary needed?

MaStr11 commented:
@jinujoseph 

> cc @CyrusNajmabadi , @MaStr11

Why me? (You probably wanted @alrz. Don't you?)

Neme12 commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,333 @@
+            else if (binaryExpression.Kind() == SyntaxKind.EqualsExpression &&
+                     binaryExpression.Left.Kind() == SyntaxKind.NumericLiteralExpression)
+            {
+                canSimplify = CanSimplifyToLengthEqualsZeroExpression(
```
why not `return` directly?

Neme12 commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,643 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInvertIf)]
+        public async Task TestCurlies1()
```
I don't see any curlies here

jinujoseph commented:
yes @MaStr11 , sorry about that :-) (it was meant for @alrz) , you are welcome to review though

chborl commented:
I addressed all of the feedback except formatting.  Still working on a good fix for that.

CyrusNajmabadi commented:
> I addressed all of the feedback except formatting. Still working on a good fix for that.

Can you explain what the issue is around formatting?  I would think that formatting would almost be a non-issue for this feature as we can just use hte existing formatting nearly entirely as-is.

Thanks!

chborl commented:
Hi @CyrusNajmabadi , when I remove formatting I still get the else on the next line because I'm moving the previous else statement (which had a newline at the end) to the if statement.  So I'm fixing that up next and updating the unit tests.

heejaechang commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,161 @@
+
+        private bool IsNumericLiteral(IOperation operation)
+        {
+            operation = RemoveImplicitConversion(operation);
```
this is not needed anymore

heejaechang commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,423 @@
+
+        Private Shared ReadOnly s_ifNodeAnnotation As New SyntaxAnnotation
+
+        Protected Shared Async Function FindIfNodeAsync(document As Document, cancellationToken As CancellationToken) As Task(Of SyntaxNode)
```
I think this can be removed now

mavasani commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,218 @@
+                        : SyntaxKind.EqualsExpression;
+                }
+
+                result = SyntaxFactory.BinaryExpression(
```
Do we handle user defined operators? Ideally, we should check if existing binary operation has a user defined operator and if so, the negated operator should also be user defined. If not, we should bail out.

mavasani commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,218 @@
+                var operation = semanticModel.GetOperation(binaryExpression);
+                if (IsSpecialCaseBinaryExpression(operation as IBinaryOperation, cancellationToken))
+                {
+                    negatedOperatorType = binaryExpression.OperatorToken.Kind() == SyntaxKind.EqualsEqualsToken
```
Can this code be base on language agnostic [BinaryOperatorKind](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/BinaryOperatorKind.cs,edd497d4cdb08e42) instead of syntax kinds? If so, we can move it down to common layer.

mavasani commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,218 @@
+                if (IsSpecialCaseBinaryExpression(operation as IBinaryOperation, cancellationToken))
+                {
+                    negatedOperatorType = binaryExpression.OperatorToken.Kind() == SyntaxKind.EqualsEqualsToken
+                        ? binaryExpression.Right is LiteralExpressionSyntax ? SyntaxKind.GreaterThanToken : SyntaxKind.LessThanToken
```
> binaryExpression.Right is LiteralExpressionSyntax

Instead of checking for literal syntax, consider checking for `IOperation.ConstantValue.HasValue`.

mavasani commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,218 @@
+                case SyntaxKind.TrueLiteralExpression:
+                {
+                    var literalExpression = (LiteralExpressionSyntax)expression;
+                    return SyntaxFactory.LiteralExpression(
```
Why not use the SyntaxGenerator helpers, such as [SyntaxGenerator.FalseLiteralExpression](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Editing/SyntaxGenerator.cs,29950d27013e228e,references)?

CyrusNajmabadi commented:
> when I remove formatting I still get the else on the next line because I'm moving the previous else statement (which had a newline at the end) to the if statement. So I'm fixing that up next and updating the unit tests.

Can you clarify this with an example?  I'm not sure i get it.  Thanks!

chborl commented:
@CyrusNajmabadi , single line ifs end up with a newline before the else, so
this:
`@"[||]if (a) { a(); } else { b(); }"`
becomes: 
`@"if (!a) { b(); }`
` else { a(); }");`

CyrusNajmabadi commented:
@chborl Thanks.  got it now.

Yeah, it seems like the trailing trivia on each of those curlies should be swapped when making this change (if it's on a single line like this).

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -445,8 +413,8 @@ void F()
         }
-        else
-        { 
+else
```
seems bad :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -53,18 +53,41 @@ protected override SyntaxNode GetIfStatement(TextSpan textSpan, SyntaxToken toke
+                // generator.IfStatement(Negate(ifNode.Condition, model, cancellationToken), ifNode.Else.Statement, ifNode.Else.WithStatement(ifNode.Statement))
+                newIfNodeStatement = ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement
+               ? SyntaxFactory.Block(ifNode.Else.Statement)
+               : ifNode.Else.Statement;
```
wonky indentation.

chborl commented:
Getting closer.  I have a couple of formatting tests to fix and other outstanding work such as, move negate to the common layer using IOperation, handle user defined operators, and other small improvements.

chborl commented:
Please review, @dotnet/roslyn-analysis and @dotnet/roslyn-ide.  All tests are passing.  

I have one item still to cover in this PR: user defined operators. @mavasani  will help me with that later today.  

Other known issues and suggested improvements are listed here: https://github.com/dotnet/roslyn/issues/25241

mavasani commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,235 @@
+        if (ifNode.Statement.GetTrailingTrivia().Any(trivia => trivia.Kind() == SyntaxKind.EndOfLineTrivia))
+        {
+            // multiline if
+            newIfNodeStatement = ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement  // if it's an else-if
```
Should this check instead be `ifNode.Else.Statement.Kind() != SyntaxKind.BlockStatment`?

mavasani commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,235 @@
+
+            newIfNodeStatement = ifNode.Else.Statement.WithTrailingTrivia(ifTrailingTrivia);
+            newElseStatement = ifNode.Else.WithStatement(ifNode.Statement).WithTrailingTrivia(elseTrailingTrivia);
+            ifNode = ifNode.WithCondition(Negate(ifNode.Condition, model, cancellationToken))
```
Can we share some code between if and else parts by moving this statement outside the if-else statement?
```
   ifNode = ifNode.WithCondition(Negate(ifNode.Condition, model, cancellationToken)) 
            .WithStatement(ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement && newIfNodeStatement.Kind() != SyntaxKind.Block ?  SyntaxFactory.Block(newIfNodeStatement) : newIfNodeStatement)
            .WithElse(newElseStatement); 
```

jinujoseph commented:
@alrz  this is in consideration for 15.7.x 
If it misses then we will take this in 15.8, based on that we can retarget your PR to the corresponding branch.

chborl commented:
Responded to feedback and added code for user defined operators.  Still need to add unit test for that.

chborl commented:
After discussion with @mavasani , it was determined we don't need the check for user defined operations because the compiler ensures the user has defined both operators.  So that section was removed and no unit test added.

CyrusNajmabadi commented:
@chborl Is this ready for another review.  Or should i wait until you ask?

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,715 @@
+            goo(); /*17*/
+            /*18*/
+        } /*19*/
+        /*20*/
```
can we have a test like this.  but for the single line case?

jinujoseph commented:
@CyrusNajmabadi this should be ready for review

chborl commented:
@CyrusNajmabadi , yes please review again.  Still planning to move negate and TryNegateBinaryComparisonExpression to the common layer using IOperation.  Thanks!

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,715 @@
+        {
+            await TestFixOneAsync(
+@"string x; [||]if (x.Length > 0) { GreaterThanZero(); } else { EqualsZero(); } } } ",
+@"string x; if (x.Length == 0) { EqualsZero(); } else { GreaterThanZero(); } } } ");
```
pretty cool :)

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,715 @@
+
+        [WorkItem(545986, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/545986")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInvertIf)]
+        public async Task TestSimplifyToLengthEqualsZero10()
```
this name is not correct.  there is no Length here.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,715 @@
+        {
+            await TestFixOneAsync(
+@"ulong x = 1; [||]if (x == 0) { a(); } else { b(); } } } ",
+@"ulong x = 1; if (x > 0) { b(); } else { a(); } } } ");
```
i don't ike this.  why did we convert ```==``` to ```>```  instead of ```!=```?

chborl replied:
Since unsigned numbers can't be negative, it can only be `> 0`.  For signed numbers you would get `!= 0` as expected.

CyrusNajmabadi replied:
I get that that translation is "correct".  I'm saying i strictly do not like it :).  If i use == on an unsigned value, i still want != when negated.  I'm *ok* with moving in the other direction.  i.e. if i have  ```if (x > 0)``` then i'm ok with you converting that to ```if (x == 0)```.  

Note: this is specifically about unsigned types.  I'm totally ok with special logic around Array/String.Length.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,715 @@
+@"string x; if (x.Length <= 0.0f) { EqualsZero(); } else { GreaterThanZero(); } } } ");
+        }
+    }
+}
```
did i miss tests with ```if { } else if { } else { }```?  i.e. 'else if' cases?

mavasani commented:
> Still planning to move negate and TryNegateBinaryComparisonExpression to the common layer using IOperation as a follow-up PR 

@chborl I presume you meant move the language specific code generation to use the `SyntaxEditor` and/or `SyntaxGenerator`. I don't think IOperation is really a requirement here.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/CodeActions/InvertIf/InvertIfTests.vb:
```diff
@@ -0,0 +1,1353 @@
+End Sub
+End Module",
+"Sub Main()
+    If (Not a OrElse Not b) And Not c Then b() Else a()
```
yeesh.  i feel like this is actually worse.  but i'm not sure hte best thing to do about it.  But ```Not (a AndAlso b Or c``` is just much nicer here.  Have we considered comparing the size of the expression before/after propogating the 'not' downwards to make sure ew pick the smaller of the two?

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/CodeActions/InvertIf/InvertIfTests.vb:
```diff
@@ -0,0 +1,1353 @@
+End Module",
+"Module Program
+    Sub Main()
+        If 0 > (<x/>.GetHashCode) Then b() Else a()
```
hrmm.. why were these parens not removed?

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,715 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInvertIf)]
+        public async Task TestMissingOnNonEmptySpan()
```
note: we probably want to allow the feature if someone exactly select the "if".

chborl commented:
@mavasani , I think we need use IOperation to get OperationKind instead of using language specific SyntaxKind, right?

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/CodeActions/InvertIf/InvertIfTests.cs:
```diff
@@ -0,0 +1,715 @@
+        {
+            await TestFixOneAsync(
+@"ulong x = 1; [||]if (0 == x) { a(); } else { b(); } } } ",
+@"ulong x = 1; if (0 < x) { b(); } else { a(); } } } ");
```
i do not like that this flipped from ```==``` to  ```<```.  I think flipping hte other direction is fine.  But i do not want my strict equality relationals becoming comparisons.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeaturesResources.Designer.cs:
```diff
@@ -548,7 +548,16 @@ internal class CSharpFeaturesResources {
+        /// </summary>
+        internal static string Invert_if_statement {
+            get {
+                return ResourceManager.GetString("Invert_if_statement", resourceCulture);
```
Note: based on current discussions around namign the "Convert 'for/query/foreach'" features, this should probably just be "Invert 'if'" (and "Invert 'If'" for VB).

CyrusNajmabadi replied:
but i don't feel strongly.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.InvertIf), Shared]
+    internal partial class InvertIfCodeRefactoringProvider : AbstractInvertIfCodeRefactoringProvider
+    {
+        private static readonly Dictionary<SyntaxKind, (SyntaxKind, SyntaxKind)> s_binaryMap =
```
i feel like the tuple shoud have member names.  knowing what Item1 and Item2 mean is not at all clear.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            return ifStatement;
+        }
+
+        protected override SyntaxNode GetRootWithInvertIfStatement(Workspace workspace, SemanticModel model, SyntaxNode ifStatement, CancellationToken cancellationToken)
```
nit, consider wrapping this long line.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+
+        protected override SyntaxNode GetRootWithInvertIfStatement(Workspace workspace, SemanticModel model, SyntaxNode ifStatement, CancellationToken cancellationToken)
+        {
+            var ifNode = (IfStatementSyntax)ifStatement;
```
Nit: our general pattern with analyzers/refactorings htat share a base class is to make the base class parameterized on these types.  That way you don't need to do these blind casts here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                // multiline if
+                newIfNodeStatement = ifNode.Else.Statement.Kind() != SyntaxKind.Block
+                ? SyntaxFactory.Block(ifNode.Else.Statement)
+                : ifNode.Else.Statement;
```
indent properly.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            var isMultiLine = ifNode.Statement.GetTrailingTrivia().Any(trivia => trivia.Kind() == SyntaxKind.EndOfLineTrivia);
+            if (isMultiLine)
+            {
+                // multiline if
```
this comment seems *very* unnecessary given the line right above :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            }
+
+            ifNode = ifNode.WithCondition(Negate(ifNode.Condition, model, cancellationToken))
+                .WithStatement(ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement && newIfNodeStatement.Kind() != SyntaxKind.Block ? SyntaxFactory.Block(newIfNodeStatement) : newIfNodeStatement)
```
please pull out hte argument to WithStatement into its own variable.  it's way too complex and confusing to read in this form.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+            }
+
+            // get new root
+            return model.SyntaxTree.GetRoot().ReplaceNode(ifStatement, ifNode);
```
is this logic hte same for the C# and VB impls?  If so, can we just pull this up into the base class, and just have the subclass responsible for providin the replacement node?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                }
+
+                case SyntaxKind.LogicalNotExpression:
+                {
```
nit: consider breaking out all these complex cases into helper methods you dispatch to.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                    var nextToken = logicalNotExpression.Operand.GetFirstToken(
+                        includeZeroWidth: true, includeSkipped: true, includeDirectives: true, includeDocumentationComments: true);
+
+                    var existingTrivia = SyntaxFactory.TriviaList(
```
i think this needs comments explaining what this is for.

chborl replied:
@CyrusNajmabadi , upon closer inspection I'm not sure this is needed.  I added a test case for `if /*0*/(/*1*/!a/*2*/)/*3*/ { a(); } else { b(); }`  and  notToken doesn't have any leading or trailing trivia.  Is there any way you know of for the not Token to have trivia?

CyrusNajmabadi replied:
sure.    ```if (!/**/a)```

chborl replied:
Ah, that's the one.  Putting trivia between the not Token and the identifier is the test case needed here.  Will add the comment and the test case.  Thanks!

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+
+                    var updatedNextToken = nextToken.WithLeadingTrivia(existingTrivia);
+
+                    return logicalNotExpression.Operand.ReplaceToken(
```
while it's pretty clear what you're doing here... i think we shoudl still have general comments like "If we're negating a ```!expr``` expression, just remove the existing ```!```).

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                }
+
+                case SyntaxKind.LogicalOrExpression:
+                {
```
1. nit: consider helper method.
2. please add comment explaining the translation.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                        Negate(binaryExpression.Right, semanticModel, cancellationToken));
+
+                    return result
+                        .Parenthesize()
```
do we have tests involving precedence changes and thus these parens shoudl/shouldn't be removed accordingly?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,238 @@
+                    SyntaxFactory.Token(
+                        binaryExpression.OperatorToken.LeadingTrivia,
+                        negatedOperatorType,
+                        binaryExpression.OperatorToken.TrailingTrivia),
```
this pattern of making a new token, but using the old trivia is repeated many times in this file.  Please extract helper.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,156 @@
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    FeaturesResources.Invert_if_statement,
```
this is not the right resource for VB.  VB should not use 'if', it should use 'If'.

CyrusNajmabadi commented on src/Features/Core/Portable/FeaturesResources.resx:
```diff
@@ -13,7 +13,7 @@
     
     ... ado.net/XML headers & schema ...
-    <resheader name="resmimetype">text/microsoft-resx</resheader>
+    <resheader name="resmimetype">text/microsoft-resx</resheader>4
```
?

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,373 @@
+            Dim invertedIfNode = FindIfNode(result.Root)
+
+            Dim explicitInvertedIfNode = Simplifier.Expand(invertedIfNode, result.Model, workspace, cancellationToken:=cancellationToken)
+            result = UpdateSemanticModel(result.Model, result.Root.ReplaceNode(invertedIfNode, explicitInvertedIfNode), cancellationToken)
```
i have no idea why you're updating the semantic model.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,373 @@
+
+        Private Function UpdateSemanticModel(model As SemanticModel, root As SyntaxNode, cancellationToken As CancellationToken) As (Model As SemanticModel, Root As SyntaxNode)
+            Dim newModel = model.Compilation.ReplaceSyntaxTree(model.SyntaxTree, root.SyntaxTree).GetSemanticModel(root.SyntaxTree)
+            Return (newModel, newModel.SyntaxTree.GetRoot(cancellationToken))
```
you do this update... and return the newModel.  but then you don't use it.  Can you explain what this is for?

CyrusNajmabadi replied:
Ok.  i see you use this below.  can you add a comment explaining why we need to go through a set of semantic-model tranformation steps?

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,373 @@
+            If s_comparisonInversesMap.TryGetValue(expression.Kind, inverses) Then
+                Dim binaryExpression = DirectCast(expression, BinaryExpressionSyntax)
+                Dim expressionType = inverses.Item1
+                Dim operatorType = inverses.Item2
```
tuples should have good names so we don't need this.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,373 @@
+            End If
+
+            Return False
+        End Function
```
this logic looks nearly identical to the C# side.  Why can't we share this code in the base class?

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,373 @@
+                Return result
+            End If
+
+            Select Case expression.Kind
```
these cases seem very similar to the C# version.  Can we extract common code to the base class by using ISyntaxFacts?

CyrusNajmabadi replied:
Note we woudl probably need to expand ISyntaxFacts a bit.  But i like tehn that that type gets more utility, and we remove more language specific stuff from our individual features.

chborl commented:
@mavasani @CyrusNajmabadi, please review.  Thanks :)

mavasani commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -266,235 +214,88 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InvertIf
-            End If
-
-            Return False
+        Friend Overrides Function GetBinaryOperation(expressionNode As SyntaxNode, semanticModel As SemanticModel) As IOperation
```
This method is not needed. You can invoke `semanticModel.GetOperation(expressionNode)` without the cast to `BinaryExpressionSyntax`.

mavasani commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -266,235 +214,88 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InvertIf
-
-            Return False
+        Friend Overrides Function GetBinaryOperation(expressionNode As SyntaxNode, semanticModel As SemanticModel) As IOperation
+            Return semanticModel.GetOperation(DirectCast(expressionNode, BinaryExpressionSyntax))
```
Please thread in the cancellation token into `SemanicModel.GetOperation` invocation.

mavasani commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -266,235 +214,88 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InvertIf
-
-            Return False
+        Friend Overrides Function IsConditionalAnd(binaryOperation As IBinaryOperation) As Boolean
+            Return binaryOperation.Syntax.ToString().Contains("AndAlso")
```
This can never be true as you are checking binary operation's syntax.

CyrusNajmabadi commented:
Will do later tonight/tomorrow morning.  Thanks!

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb:
```diff
@@ -1573,6 +1594,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return DirectCast(node, UnaryExpressionSyntax).Operand
         End Function
 
+        Public Function GetOperatorTokeOfPrefixUnaryExpression(node As SyntaxNode) As SyntaxToken Implements ISyntaxFactsService.GetOperatorTokenOfPrefixUnaryExpression
```
"Token" not "Toke"

CyrusNajmabadi commented on src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb:
```diff
@@ -662,6 +662,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+        End Function
+
+        Public Function IsIfStatement(node As SyntaxNode) As Boolean Implements ISyntaxFactsService.IsIfStatement
+            Return (node.Kind() = SyntaxKind.IfStatement)
```
unnecessary parens.

CyrusNajmabadi replied:
still applies.

CyrusNajmabadi replied:
still applies.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs:
```diff
@@ -78,6 +78,8 @@ internal interface ISyntaxFactsService : ILanguageService
 
         bool IsBinaryExpression(SyntaxNode node);
         void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right);
+        SyntaxToken GetOperatorTokenOfBinaryExpression(SyntaxNode node);
```
we'll conflict.  that's ok.  :)

CyrusNajmabadi replied:
i have similar changes in the 'require parentheses' PR. :)

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -728,6 +728,21 @@ public SyntaxNode GetNameOfAttribute(SyntaxNode node)
+        public bool IsIfStatement(SyntaxNode node)
+        {
+            return (node.Kind() == SyntaxKind.IfStatement);
+        }
```
unnecessary parens.

nit, use=> for all of these.

CyrusNajmabadi replied:
this still applies.

CyrusNajmabadi replied:
this still applies.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
-                base(title, createChangedDocument)
-            {
-            }
+            return ((LiteralExpressionSyntax)expression).IsKind(SyntaxKind.TrueLiteralExpression);
```
1. consider following the pattern with other refactoring/fixproviders, and make the base type generic and parametrized on these types.  That way you don't need all these casts in your code.
2. you don't need the cast at all, you can just do ```expression.IsKind(...)```
3. You could just make these SyntaxFacts.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
+        internal override bool IsConditionalOr(SyntaxToken operatorToken)
+        {
+            return operatorToken.Kind() == SyntaxKind.BarBarToken;
+        }
```
syntaxfact?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
         }
 
-        private ExpressionSyntax Negate(ExpressionSyntax expression, SemanticModel semanticModel, CancellationToken cancellationToken)
+        internal override bool IsConditionalAnd(SyntaxToken operatorToken)
```
do you need to check the token?  wouldn't you have a binary expression to begin with that you could check?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
+        internal override string GetInvertIfText()
+        {
+            return CSharpFeaturesResources.Invert_if;
         }
```
nit: =>

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
-            SemanticModel semanticModel,
-            CancellationToken cancellationToken,
-            out ExpressionSyntax result)
+        private void GetPartsOfIfStatement(SyntaxNode node, out SyntaxNode condition, out SyntaxNode thenStatements, out SyntaxNode elseStatements)
```
i cannot find where this is called.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
 
-            return false;
+            // get new root
+            return model.SyntaxTree.GetRoot().ReplaceNode(ifStatement, ifNode);
```
GetRoot should take the cancellation token.

CyrusNajmabadi replied:
Why not just return the new if statement instead of updating the root?  The caller can update the root themselves.  Now this type is just responsible for figuring out how to do the inversion.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
+            var syntaxFacts = GetSyntaxFactsService();
 
-            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var ifNode = (IfStatementSyntax)ifStatement;
```
ifStatement and ifNode are not good names.  perhaps oldIfStatement and newIfStatement?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
-            }
-            else if (binaryExpression.Kind() == SyntaxKind.EqualsExpression &&
-                     binaryExpression.Right.Kind() == SyntaxKind.NumericLiteralExpression)
+            var isMultiLine = ifNode.Statement.GetTrailingTrivia().Any(trivia => trivia.Kind() == SyntaxKind.EndOfLineTrivia);
```
what if you have:

```c#
if (foo) {
   x();
   x();
} else {
   y();
   y();
}
```

That code would say this is not multiline.  Is that intentional?

chborl replied:
You're right.  I changed the variable to `hasNewLineAfterClosingBrace` because it does not catch all types of multiline statements.  I also added the above code to a test and it passes.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
-                var symbol = semanticModel.GetSymbolInfo(variableExpression, cancellationToken).Symbol;
+            var newIfStatment = ifNode.Else.Statement.Kind() == SyntaxKind.IfStatement && newIfNodeStatement.Kind() != SyntaxKind.Block
+                        ? SyntaxFactory.Block(newIfNodeStatement)
+                        : newIfNodeStatement;
```
indented too far. 
"Statement" not "Statment"

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
-                }
+            ifNode = ifNode.WithCondition((ExpressionSyntax)(Negate(ifNode.Condition, generator, syntaxFacts, model, cancellationToken)))
+                .WithStatement(newIfStatment)
+                .WithElse(newElseStatement);
```
see comments below about how you can do this without needing all the explicit casts here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
         }
 
-        private async Task<Document> InvertIfAsync(Document document, IfStatementSyntax ifStatement, CancellationToken cancellationToken)
+        internal override SyntaxNode GetRootWithInvertIfStatement(
```
does this need to be internal?  why not just protected?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,123 @@
-    // [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.InvertIf)]
-    internal partial class InvertIfCodeRefactoringProvider : CodeRefactoringProvider
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = PredefinedCodeRefactoringProviderNames.InvertIf), Shared]
+    internal partial class InvertIfCodeRefactoringProvider : AbstractInvertIfCodeRefactoringProvider
```
This should be called CSharpInvertIf... That's our naming for providers with inheritance.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+        internal abstract bool IsFalseLiteralExpression(SyntaxNode expression);
+        internal abstract bool IsTrueLiteralExpression(SyntaxNode expression);
+
+
```
extra blank line

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+        internal abstract bool IsConditionalAnd(SyntaxToken operatorToken);
+        internal abstract bool IsConditionalOr(SyntaxToken operatorToken);
+        internal abstract bool IsFalseLiteralExpression(SyntaxNode expression);
+        internal abstract bool IsTrueLiteralExpression(SyntaxNode expression);
```
why internal and not protected?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+
+            context.RegisterRefactoring(
+                new MyCodeAction(
+                    GetInvertIfText(),
```
usually we call this GetTitle()  or Title

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            return document.WithSyntaxRoot(
+                // this returns the root because VB requires changing context around if statement
```
ah, ignore my question about jusst returning the if-statement.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+    SyntaxGenerator generator,
+    ISyntaxFactsService syntaxFacts,
+    SemanticModel semanticModel,
+    CancellationToken cancellationToken)
```
indentation busted.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+            {
+                return (GetNegationOfLogicalNotExpression(expression, generator, syntaxFacts, semanticModel));
+            }
+            return generator.LogicalNotExpression(expression);
```
nit: newline above this.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+            CancellationToken cancellationToken)
+        {
+            syntaxFacts.GetPartsOfBinaryExpression(expressionNode, out var leftOperand, out var rightOperand);
+            var operatorToken = syntaxFacts.GetOperatorTokenOfBinaryExpression(expressionNode);
```
nit: just have GetPartsOfBinaryExpression also include the token.  Then add a GetPartsOfBinaryExpression extension method to ISyntaxFactsExtensions that keeps hte old overload and forwards it to the new method.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+            {
+                return generator.LogicalNotExpression(expressionNode);
+            }
+            var binaryOperation = (IBinaryOperation)operation;
```
nit, newline after }

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+                {
+                    newLeftOperand = leftOperand;
+                    newRightOperand = rightOperand;
+                }
```
consider:

``
var newLeftOperand = leftOperand;
var newRightOperand = rightOperand;
if (negateOperands) {
   ...
}
```

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+                    else
+                    {
+                        return generator.ReferenceEqualsExpression(leftOperand, rightOperand);
+                    }
```
nit: use ternary expression:   i.e. ```return ... ? ... : ...```

CyrusNajmabadi replied:
you fixed up all the rest *except* this one :)

chborl replied:
doh!  fixed it now :)

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+                    return generator.LogicalOrExpression(leftOperand, rightOperand);
+                case BinaryOperatorKind.ConditionalAnd:
+                    return generator.LogicalAndExpression(leftOperand, rightOperand);
+            }
```
blank line after }

CyrusNajmabadi replied:
you added a blank line after the "return null" statement not after the } :)

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+                return RemoveImplicitConversion(conversion.Operand);
+            }
+
+            return operation;
```
nit: use => and ternary.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+private SyntaxNode GetNegationOfLiteralExpression(
+    SyntaxNode expression,
+    SyntaxGenerator generator,
+    SemanticModel semanticModel)
```
busted indentation.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+            {
+                newLiteralExpression = generator.LogicalNotExpression(expression);
+            }
+            return newLiteralExpression.WithLeadingTrivia(leadingExpressionTrivia).WithTrailingTrivia(trailingExpressionTrivia);
```
you can just say ```newLiteralExpression.WithTriviaFrom(expression)```

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+            }
+            else
+            {
+                newLiteralExpression = generator.LogicalNotExpression(expression);
```
this seems wrong.  if 'expression' has trivia on it, you'll wrap that trivia, then add it on again afterwards because of the code below.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+    SyntaxNode expression,
+    SyntaxGenerator generator,
+    ISyntaxFactsService syntaxFacts,
+    SemanticModel semanticModel)
```
indentation.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,392 @@
+    SyntaxNode expression,
+    SyntaxGenerator generator,
+    ISyntaxFactsService syntaxFacts,
+    SemanticModel semanticModel)
```
you don't use semanticModel.

CyrusNajmabadi replied:
you don't use generator

CyrusNajmabadi commented:
Done with this pass.  When this stuff is fixed up let me know and i can take another look.  Also, i have not reviewed the VB side yet.

chborl commented:
Many thanks for the review, @CyrusNajmabadi! 

I have addressed all of your feedback except for the following, which I will do first thing tomorrow:
- make base type generic and parameterize to avoid direct casts
- look into the multiline issue you mentioned.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract string GetTitle();
+        protected abstract bool IsFalseLiteralExpression(SyntaxNode expression);
+        protected abstract bool IsTrueLiteralExpression(SyntaxNode expression);
```
can;t be syntax factss?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+            {
+                return syntaxFacts.Parenthesize(
+                    Negate(
+                        syntaxFacts.GetExpressionOfParenthesizedExpression(expression),
```
this expression likely need to have trivia stripped.

chborl replied:
Can you elaborate why the trivia should be stripped from the expression?  I added new tests with trivia around parentheses and other than this new issue, https://github.com/dotnet/roslyn/issues/26427, didn't have problems.

CyrusNajmabadi replied:
ah, sorry.  i misread this.  i thought this was just negating the expression, and then adding the expression's trivia to the outside negation.  i see now that it's grabbing the expression *inside* the parens.  So this looks fine to me.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+                        semanticModel,
+                        cancellationToken))
+                    .WithLeadingTrivia(expression.GetLeadingTrivia())
+                    .WithTrailingTrivia(expression.GetTrailingTrivia());
```
can just say .WithTriviaFrom(expression).

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+            var leadingTokenTrivia = operatorToken.LeadingTrivia;
+            var trailingTokenTrivia = operatorToken.TrailingTrivia;
+            var leadingExpressionTrivia = expressionNode.GetLeadingTrivia();
+            var trailingExpressionTrivia = expressionNode.GetTrailingTrivia();
```
dont need these four variables

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+
+                var newBinaryExpressionSyntax = NewBinaryOperation(binaryOperation, newLeftOperand, negatedKind, newRightOperand, generator, syntaxFacts, cancellationToken)
+                    .WithLeadingTrivia(leadingExpressionTrivia)
+                    .WithTrailingTrivia(trailingExpressionTrivia);
```
can use just WithTriviaFrom

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+                    {
+                        var containingType = property.ContainingType;
+                        if (containingType != null &&
+                            (containingType.SpecialType == SpecialType.System_Array ||
```
could remove null check and just do: ```containingType?.SpecialType == System_Array || ...```

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+                        newLiteralExpression = generator.LogicalNotExpression(expression.WithoutLeadingTrivia().WithoutTrailingTrivia());
+                    }
+                    return newLiteralExpression.WithTriviaFrom(expression);
+                }
```
dedent whole method.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+                    {
+                        newLiteralExpression = generator.LogicalNotExpression(expression.WithoutLeadingTrivia().WithoutTrailingTrivia());
+                    }
+                    return newLiteralExpression.WithTriviaFrom(expression);
```
newline after } plz.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,363 @@
+                    }
+                    else
+                    {
+                        newLiteralExpression = generator.LogicalNotExpression(expression.WithoutLeadingTrivia().WithoutTrailingTrivia());
```
can just use expression.WithoutTrivia()

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -1613,6 +1628,15 @@ public void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out
 
+        public void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right, out SyntaxToken token)
+        {
+            GetPartsOfBinaryExpression(node, out left, out right);
```
i would do things the reverse way.  have the Simple version call into this version.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -1627,12 +1651,25 @@ public SyntaxNode WalkDownParentheses(SyntaxNode node)
+        public bool IsConditionalOr(SyntaxNode node)
+            => node.Kind() == SyntaxKind.LogicalOrExpression;
+
+
```
extra newline.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs:
```diff
@@ -78,6 +78,9 @@ internal interface ISyntaxFactsService : ILanguageService
 
         bool IsBinaryExpression(SyntaxNode node);
         void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right);
```
this can be moved to be in ISyntaxFactsExtensions now.

chborl replied:
How do we determine when to move a method to extensions?

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs:
```diff
@@ -78,6 +78,9 @@ internal interface ISyntaxFactsService : ILanguageService
         bool IsBinaryExpression(SyntaxNode node);
         void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right);
+        void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right, out SyntaxToken token);
+        SyntaxToken GetOperatorTokenOfBinaryExpression(SyntaxNode node);
```
this can  also be an extension.

CyrusNajmabadi commented:
Note: 

> make base type generic and parameterize to avoid direct casts

i don't think you have to do this.  Most of your calls will just become syntax facts calls, so i think it's sufficient.  That will also be much easier than refactoring this to go the generic route.  but it's something to consider for the future!

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -1,177 +1,124 @@
 
                 relevantIfBlockOrIfStatement = singleLineIf
-                relevantSpan = singleLineIf.IfKeyword.Span
+                relevantSpan = singleLineIf.Span
```
this could be a pretty large span.  perhaps limit it just to the if+conditoin.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -1,177 +1,124 @@
-                        Return
-                    End If
-                End If
+                relevantSpan = multiLineIf.IfStatement.IfKeyword.Span
```
this is different than C#, intentional?

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -1,177 +1,124 @@
-                Return
+            If token.SyntaxTree.OverlapsHiddenPosition(relevantSpan, cancellationToken) Then
+                Return Nothing
             End If
```
do we also have a check that we don't offer the feature if the if-node itself spans anything hidden?  if so, we don't really need this check.  (and we should have the total-span check in the base type).

CyrusNajmabadi replied:
not needed here right?  yo uhave the check in the base class.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -1,177 +1,124 @@
-            ifNode = Await FindIfNodeAsync(_updatedDocument, cancellationToken).ConfigureAwait(False)
+            Dim ifNode = FindIfNode(result.Root)
 
             ' Complexify the top-most statement parenting this if-statement if necessary
```
can yuo explain why we need this for VB, but not for C#?

chborl replied:
This is needed for the following 2 tests in VB:
1)  `Dim x = Sub() If True Then Dim y Else Console.WriteLine(), z = 1` inverts to 
    `Dim x = (Sub() If False Then Console.WriteLine() Else Dim y), z = 1`
2) `Case Sub() If True Then Dim x Else Return, Nothing` inverts to
     `Case (Sub() If False Then Return Else Dim x), Nothing`

I'm trying to figure out if there is an equivalent in C# that also needs expanded.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -1,177 +1,124 @@
+            Return root _
+                       .GetAnnotatedNodesAndTokens(s_ifNodeAnnotation) _
+                       .Single() _
+                       .AsNode()
```
nit: i feel like this method could be a single line, and then inlined.  

your call.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -1,177 +1,124 @@
+            Return (newModel, newModel.SyntaxTree.GetRoot(cancellationToken))
+        End Function
+
+        Private Shared ReadOnly s_ifNodeAnnotation As New SyntaxAnnotation
```
move to top of class.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -266,235 +214,74 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InvertIf
+            Return result.Model
+        End Function
+
+        Private Function GetIfStatementInfo(node As SyntaxNode, condition As SyntaxNode, statements As SyntaxNode, elseIfStatements As SyntaxNode, elseStatements As SyntaxNode, IsSingleLineStatement As Boolean) As Boolean
```
i see no callers to this...

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -266,235 +214,74 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.InvertIf
-
-            Return result.WithAdditionalAnnotations(Formatter.Annotation)
+        Protected Overrides Function IsTrueLiteralExpression(expression As SyntaxNode) As Boolean
+            Return DirectCast(expression, LiteralExpressionSyntax).IsKind(SyntaxKind.TrueLiteralExpression)
```
don't need the cast.  also, should be a SyntaxFacts.

CyrusNajmabadi commented:
> This is needed for the following 2 tests in VB:

```vb
    Dim x = Sub() If True Then Dim y Else Console.WriteLine(), z = 1 inverts to
    Dim x = (Sub() If False Then Console.WriteLine() Else Dim y), z = 1
```

That is an epic test case.

Since we have the fix for it, i'm ok with the fixing staying.  that said... i would also be ok with us just making things simple and saying "will not invert single-line ifs in single-line lambdas.

That said, let's keep for now.  But a comment explaining this case in the code would be good.

CyrusNajmabadi commented:
@chborl A similar case for c# might be:

```c#
if (foo) 
   bar();
else { 
   if (baz)
      Quux();
}
```

A naive inversion will conver the above to:

```c#
if (!foo) 
   if (baz)
      Quux();
else { 
   bar();
}
```

However, this will change meaning as hte 'else' will now be parsed as part of the inner 'if' instead of hte outer 'if'.

Do we handle this case and do we have a test for this?

chborl commented:
Most feedback addressed, but will follow up on the following tomorrow:
- move sytnaxfacts to extensions
- change trigger for C# and VB
- spans for hidden
- add comment for the VB code section (I already added the test case for C# and it was passing)

chborl commented:
Just pushed the latest.  Will address the rest either later tonight or tomorrow morning.  Thanks!

chborl commented:
@CyrusNajmabadi , can you take another look?  I've addressed the rest of the feedback and opened a new issue to not offer invert if in C# lambda expressions.

CyrusNajmabadi commented:
> And opened a new issue to not offer invert if in C# lambda expressions.

Do you mean in 'vb' lambdas?  C# should be no problem, right?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.cs:
```diff
@@ -1,332 +1,93 @@
             }
 
-            if (!ifStatement.IfKeyword.Span.IntersectsWith(textSpan.Start))
+            var span = TextSpan.FromBounds(ifStatement.IfKeyword.Span.Start, ifStatement.Condition.Span.End);
```
this is wacky for C#.   you are selecting this span:  ```[|if (foo|])```.  That does not seem to be right at all.

chborl replied:
Since I limited VB to only span If + Condition I thought I should be consistent with C#.  If not, then what would make the most sense for C#?

CyrusNajmabadi replied:
probably ```[|if (foo)|]```

CyrusNajmabadi commented on src/Features/Core/Portable/CodeRefactorings/InvertIf/AbstractInvertIfCodeRefactoringProvider.cs:
```diff
@@ -0,0 +1,350 @@
+            {BinaryOperatorKind.And, BinaryOperatorKind.Or},
+            {BinaryOperatorKind.ConditionalOr, BinaryOperatorKind.ConditionalAnd},
+            {BinaryOperatorKind.ConditionalAnd, BinaryOperatorKind.ConditionalOr},
+        };
```
why do you need this?

chborl replied:
I'm using it to get the opposite operator of a given binary operator.  This could also be done in a method with a select statement.  Or, any other suggestions?

CyrusNajmabadi replied:
Ignore my comment. I thought I was looking at a separate file

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs:
```diff
@@ -77,7 +80,9 @@ internal interface ISyntaxFactsService : ILanguageService
 
         bool IsBinaryExpression(SyntaxNode node);
-        void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right);
+        void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right, out SyntaxToken token);
```
nit: the operaotr token should be between left and right.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs:
```diff
@@ -77,7 +80,9 @@ internal interface ISyntaxFactsService : ILanguageService
         bool IsBinaryExpression(SyntaxNode node);
-        void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right);
+        void GetPartsOfBinaryExpression(SyntaxNode node, out SyntaxNode left, out SyntaxNode right, out SyntaxToken token);
+        SyntaxToken GetOperatorTokenOfBinaryExpression(SyntaxNode node);
```
this can just be an extension method.  no need for c# or VB to impl it, you can just build it off the GetPartsOfBinaryExpression method.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -1808,6 +1839,12 @@ public SyntaxNode WithModifiers(SyntaxNode node, SyntaxTokenList modifiers)
+            => ((LiteralExpressionSyntax)expression).IsKind(SyntaxKind.FalseLiteralExpression);
+
+        public bool IsTrueLiteralExpression(SyntaxNode expression)
+            => ((LiteralExpressionSyntax)expression).IsKind(SyntaxKind.TrueLiteralExpression);
```
don't need the casts.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -1606,13 +1621,17 @@ public bool IsDefaultLiteralExpression(SyntaxNode node)
         }
 
+        public SyntaxToken GetOperatorTokenOfBinaryExpression(SyntaxNode node)
+            => ((BinaryExpressionSyntax)node).OperatorToken;
```
uneeded, this could be an extension method.

CyrusNajmabadi replied:
you made this an extension method.  so this method should go away.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -728,6 +728,21 @@ public SyntaxNode GetNameOfAttribute(SyntaxNode node)
 
+        public bool IsParenthesizedExpression(SyntaxNode node)
+        {
+            return (node.Kind() == SyntaxKind.ParenthesizedExpression);
```
unnecessary parens.

chborl commented:
>Do you mean in 'vb' lambdas? C# should be no problem, right?

I must've misunderstood.  For VB lambdas, the fix is already in the code.  But there isn't an equivalent fix in C# so I thought that's what you meant by  'i would also be ok with us just making things simple and saying "will not invert single-line ifs in single-line lambdas." '  I was going to write a similar test in C# (if that's even possible, I don't know) and then add the check to not offer the fix in that case.  

If you don't think that's necessary, or even possible, for C# then I'll close #26454.

CyrusNajmabadi commented:
This still needs to be answered: https://github.com/dotnet/roslyn/pull/25233#issuecomment-384812525

CyrusNajmabadi commented:
> I must've misunderstood. For VB lambdas, the fix is already in the code. But there isn't an equivalent fix in C# so I thought that's what you meant by 'i would also be ok with us just making things simple and saying "will not invert single-line ifs in single-line lambdas." ' I was going to write a similar test in C# (if that's even possible, I don't know) and then add the check to not offer the fix in that case.

I don't think C# has ths problem.

I was basically saying: i would be ok *removing* the code VB has and having it just punt on this case since it's so marginal.  IN other words, not offer 'invert if on a single line if statement in a single line VB sub/lambda'.  However, as i originally mentioned, because the fix was already made, might as well leave it in.

Neme12 commented:
```c#
    Dim x = Sub() If True Then Dim y Else Console.WriteLine(), z = 1 inverts to
    Dim x = (Sub() If False Then Console.WriteLine() Else Dim y), z = 1
```
I feel like these sorts of things should be handled by the simplifier.

chborl commented:
>This still needs to be answered: #25233 (comment)
>>Do we handle this case and do we have a test for this?

Yes.  I added this test case and it passed without any changes.

chborl commented:
>@Neme12 , I feel like these sorts of things should be handled by the simplifier.

Yes, it's the call to `Simplifier.Expand()` that added the parentheses.

CyrusNajmabadi commented:
@neme12 the main problem here is that the transformation ends up producing code that would be *parsed* differently. That needs to be avoided in the first place, otherwise the simplifier is effectively just operating on a busted to that doesn't reflect what the compiler would actually produce

CyrusNajmabadi commented:
The way this is handled is generally to get the simplifier to actually expand things out to make the parse clear, then have that simplify in the end to the extent possible

Neme12 commented:
Yes, but the idea is we would always add parentheses with a simplifier annotation that the simplifier removes if it finds they are unnecessary, just as this works with casts.

Neme12 commented:
I'm sorry, I haven't actually read all the code. That's great if that's what it does. Thanks

jinujoseph commented:
@chborl can you retarget this to master pls.

chborl commented:
Squashed commits to remove the WIP tag, will merge once all tests pass.

dpoeschl commented:
retest windows_debug_spanish_unit32_prtest please

dpoeschl commented:
@jaredpar Spanish failed with an AV again: https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_spanish_unit32_prtest/689/artifact/Binaries/Debug/Logs/Roslyn.ExpressionEvaluator.VisualBasic.ResultProvider.UnitTests.dll.out.log

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,250 @@
+
+            If token.SyntaxTree.OverlapsHiddenPosition(relevantSpan, cancellationToken) Then
+                Return Nothing
+            End If
```
why do you need this check?  i still feel ike it's unnecessary given the similar check in the base class.

CyrusNajmabadi commented on src/Features/VisualBasic/Portable/CodeRefactorings/InvertIf/InvertIfCodeRefactoringProvider.vb:
```diff
@@ -0,0 +1,250 @@
+                         .WithElseClause(elseClause.WithStatements(ifNode.Statements).WithTrailingTrivia(elseClause.GetTrailingTrivia()))
+        End Function
+
+#If False Then
```
why is this code here?

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsService.cs:
```diff
@@ -60,6 +60,9 @@ internal interface ISyntaxFactsService : ILanguageService
+        bool IsFalseLiteralExpression(SyntaxNode expression);
+        bool IsTrueLiteralExpression(SyntaxNode expression);
+
 
```
extra newline.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/LanguageServices/SyntaxFactsService/ISyntaxFactsServiceExtensions.cs:
```diff
@@ -46,5 +46,15 @@ public static bool SpansPreprocessorDirective(this ISyntaxFactsService service,
+        public static void GetOperatorTokenOfBinaryExpression(this ISyntaxFactsService syntaxFacts, SyntaxNode node, out SyntaxToken token)
+        {
+            syntaxFacts.GetPartsOfBinaryExpression(node, out _, out token, out _);
+        }
```
nit: use => for these.

alrz replied:
@CyrusNajmabadi Shouldn't we just prefer `=>` when the expression fits in the front of member? e.g. `M() => false;` otherwise IMO that's actually harder to read.

CyrusNajmabadi replied:
We've used => for multiline methods for a while now.  It doesn't seem to have a problem. It also helps signify that these methods should try to stay small and not do too much.

alrz commented:
@CyrusNajmabadi @chborl 

I'm intergaring https://github.com/dotnet/roslyn/pull/24649 on top of this change and had to do some refactorings and clean up. Still VB part is not improved much because I don't understand what's going on. Can someone explain why VB needs semantic update for this? comments just say that what we're doing, not why. also I don't understand what "complexify" means.

alrz commented:
>TestTryToParenthesizeAwkwardSyntaxInsideSingleLineLambdaMethod

Could we say we always parenthesize a singleline-if if it's inside a lambda?

alrz commented:
Or else, we could look for an exact match: "single-line `If` in a lambda in a initializer"

`Dim x = Sub() If True Then Dim y Else Console.WriteLine(), z = 1`


Honestly that code doesn't look like very frequent pattern to "complexify" the whole feature

heejaechang commented:
@alrz are you saying the test case is the only case where "inverting if" can change semantic in VB? 

also, any fixer that moves around code should always do complexify => simplify by default, and not do that only if it is confident that tree transformation is safe from semantic changes.

since VB is quite complex, no one is confident enough to list all the possible cases where inverting if can cause semantic changes, so it is doing the default behavior (safer way). if you can list all possible cases, we can remove those and do what C# does.

alrz commented:
I'm yet to see just another instance of this ambiguity, the test case seems to be targeted on a very specific pattern. I'll keep the code for now, we can discuss further in that PR to simplify things.
